pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -1241,9 +1322,71 @@ public DatabaseMetaData getMetaData() throws SQLException {
     return metadata;
   }
 
+  /**
+   * @see  #setTypeMap(java.util.Map)
+   */
+  private static Set<Class<?>> getAllClasses(Class<?> clazz) {
+    Set<Class<?>> classes = new HashSet<Class<?>>();
+    Class<?> current = clazz;
+    do {
+      classes.add(current);
+      for (Class<?> iface : clazz.getInterfaces()) {
+        classes.add(iface);
+      }
+    } while ((current = current.getSuperclass()) != null);
+    return classes;
+  }
+
+  /**
+   * @see  #setTypeMap(java.util.Map)
+   */
+  private static void addInference(Map<Class<?>, Set<String>> inferenceMap, Class<?> clazz, String type) {
+    Set<String> types = inferenceMap.get(clazz);
+    if (types == null) {
+      types = new HashSet<>();
+      inferenceMap.put(clazz, types);
+    }
+    types.add(type);
+  }
+
+  /**
+   * @see  #setTypeMap(java.util.Map)
+   */
+  private static void makeValuesUnmodifiable(Map<Class<?>, Set<String>> inferenceMap) {
+    for (Map.Entry<Class<?>, Set<String>> entry : inferenceMap.entrySet()) {
+      Set<String> types = entry.getValue();
+      if (types.size() == 1) {
+        entry.setValue(Collections.singleton(types.iterator().next()));
+      } else {
+        entry.setValue(Collections.unmodifiableSet(types));
+      }
+    }
+  }
+
   @Override
   public void setTypeMap(Map<String, Class<?>> map) throws SQLException {
-    setTypeMapImpl(map);
+    // Defensive copy
+    Map<String, Class<?>> newMap = new HashMap<>(map);
+    // Build the new direct inference map
+    Map<Class<?>, Set<String>> newDirectMap = new HashMap<>();
+    // Build the new type inference map, including all base classes and implemented interfaces
+    Map<Class<?>, Set<String>> newInheritedMap = new HashMap<>();
+    for (Map.Entry<String, Class<?>> entry : newMap.entrySet()) {
+      String type = entry.getKey();
+      Class<?> directClass = entry.getValue();
+      addInference(newDirectMap, directClass, type);
+      for (Class<?> clazz : getAllClasses(directClass)) {
+        addInference(newInheritedMap, clazz, type);
+      }
+    }
+    // Make each type set unmodifiable
+    makeValuesUnmodifiable(newDirectMap);
+    makeValuesUnmodifiable(newInheritedMap);
+    synchronized (typemapLock) {
```

Comment:
what is the purpose of this synchronization? the corresponding reads (from get methods) are not synchronized. 

Is the concern trying to handle concurrent calls to setTypeMap on the same PgConnection?

Reply:
I agree with @bokken : the synchronization should be either in both reads and writes or it should be removed completely.
It makes little sense to synchronize writes only.